frontend/expr_inferrer.py

```diff
@@ -124,6 +128,14 @@ def _infer_add(left_type, right_type, lineno, solver):
     result_type = solver.new_z3_const("addition_result")
     solver.add(axioms.add(left_type, right_type, result_type, solver.z3_types),
                fail_message="Addition in line {}".format(lineno))
+
+    # Add soft constraint in case of list addition
```

Comment:
Yes but the soft constraints are easily satisfied and they seem to speed things up. Why do you need to add the condition that the operand is a number? That is what the hard constraints do already.